src/dataflow/src/render/filter.rs

```diff
@@ -0,0 +1,273 @@
+                                        .into_iter()
+                                        .chain(None.into_iter());
+                                }
+                                Ok(Datum::True) => true,
+                                _ => false,
+                            }
+                    }
+                    if passed {
+                        // In order to work with times, it is easiest to convert it to an i128.
+                        // This is because our decimal type uses that representation, and going
+                        // from i128 to u64 is even more painful.
+                        let mut time_128 = i128::from(time);
+                        // Track whether we have seen a null in either bound, as this should
+                        // prevent the record from being produced at any time.
+                        let mut null_eval = false;
+
+                        // If there is a lower bound, advance `time` to that lower bound.
+                        // This decimal stuff is brittle; let's hope the scale never changes.
+                        for l in lower.iter() {
+                            match l.eval(&datums_local, &temp_storage) {
+                                Err(e) => {
+                                    return Some(Err((DataflowError::from(e), time, diff)))
+                                        .into_iter()
+                                        .chain(None.into_iter());
+                                }
+                                Ok(Datum::Decimal(s)) => {
+                                    if time_128 < s.as_i128() {
+                                        time_128 = s.as_i128();
+                                    }
+                                }
+                                Ok(Datum::Null) => {
+                                    null_eval = true;
+                                }
+                                x => {
+                                    panic!("Non-decimal value in temporal predicate: {:?}", x);
+                                }
+                            }
+                        }
+
+                        // Extract an minimum upper bound. If not for errors, we would just write:
+                        // let upper_bound = upper.iter().map(|l| l.eval(&datums_local, &temp_storage)?).min();
+                        let mut upper_bound = None;
+                        for u in upper.iter() {
+                            match u.eval(&datums_local, &temp_storage) {
+                                Err(e) => {
+                                    return Some(Err((DataflowError::from(e), time, diff)))
+                                        .into_iter()
+                                        .chain(None.into_iter());
+                                }
+                                Ok(Datum::Decimal(s)) => {
+                                    if upper_bound < Some(s.as_i128()) {
```

Comment:
wait hang on are we taking the minimum upper bound here or the maximum? like we get upper bounds [7, 5] then ...

none is less than 7, so the upper bound becomes 7
then 7 is not less than 5, so we keep 7 as the upper bound (talking through this myself)

it seems like this is maybe a comments bug (I'm not sure whether we actually want the min or max upper bound)

Reply:
I believe I have screwed up the logic here.